### PR TITLE
PLT-8659 UPLC term order evaluator takes into account builtin argument saturation

### PR DESCRIPTION
**Pull request by Unisay**
This PR extends UPLC term order evaluator to make a better judgement about a built-in term purity based on its  saturation with arguments. In order to make it possible I had to pass `BuiltinSemanticsVariant` all the way from the top down to evaluator.

### CODE REVIEW

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Cse.hs [hunk 1]
```diff
@@ -277,51 +266,51 @@ annotate = flip evalState 0 . flip runReaderT [] . go
...
+    -> HashMap (Term name uni fun ()) [(Path, Term name uni fun (Path, ann), Int)]
+    -> HashMap (Term name uni fun ()) [(Path, Term name uni fun (Path, ann), Int)]
+  addToMap t0
+    -- We don't consider work-free terms for CSE, because doing so may or may not
+    -- have a size benefit, but certainly doesn't have any cost benefit (the cost
+    -- will in fact be slightly higher due to the additional application).
+    --
+    -- `isWorkFree` currently doesn't check whether a builtin application is saturated,
+    -- or whether an term is the (possibly repeated) forcing of a builtin (which should
+    -- be workfree), so we check it separately.
```
Comment (michaelpj) [inline plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Cse.hs:291]: I think this comment is now inaccurate, and you can get rid of the extra conditions?

File: plutus-tx-plugin/test/AsData/Budget/9.6/patternMatching.uplc.golden [hunk 2]
```diff
@@ -3,43 +3,33 @@ program
...
-                            (addInteger cse))
-                         (case cse [(\x y z w -> x)]))
-                      (case cse [(\x y z w -> y)]))
-                   (case cse [(\x y z w -> z)]))
-                (case cse [(\x y z w -> w)]))
-             (\x y -> addInteger x y))
+                       addInteger
+                         (addInteger
+                            (addInteger
+                               (addInteger (addInteger cse cse) cse)
```
Comment (michaelpj) [inline plutus-tx-plugin/test/AsData/Budget/9.6/patternMatching.uplc.golden:13]: This seems like what we expect: `addInteger cse` is work-free, and based on the comments in CSE, we don't want to CSE such things.

**Review by michaelpj — APPROVED** [review]

File: plutus-core/executables/pir/Main.hs [hunk 3]
```diff
@@ -169,15 +169,19 @@ compileToUplc optimise plcProg =
...
             if optimise
             then PLC.defaultCompilationOpts
-            else PLC.defaultCompilationOpts & PLC.coSimplifyOpts . UPLC.soMaxSimplifierIterations .~ 0
-    in flip runReader plcCompilerOpts $ runQuoteT $ PLC.compileProgram plcProg
+            else PLC.defaultCompilationOpts
+                    & PLC.coSimplifyOpts . UPLC.soMaxSimplifierIterations .~ 0
+        defaultBuiltinSemanticsVariant :: BuiltinSemanticsVariant UPLC.DefaultFun
+        defaultBuiltinSemanticsVariant = def
+    in flip runReader plcCompilerOpts $ runQuoteT $
+         PLC.compileProgram defaultBuiltinSemanticsVariant plcProg
```
[the remark this answers is not in the document]
Comment (Unisay) [reply plutus-core/executables/pir/Main.hs]: Done

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Cse.hs [hunk 4]
```diff
@@ -277,51 +266,51 @@ annotate = flip evalState 0 . flip runReaderT [] . go
...
-        | (Builtin _ fun, args) <- splitApplication t ->
-            length args >= Map.findWithDefault 0 fun arityInfo
-      _ -> True
-
-    isForcingBuiltin = \case
-      Builtin{} -> True
-      Force _ t -> isForcingBuiltin t
-      _ -> False
+countOccs builtinSemanticsVariant = foldrOf termSubtermsDeep addToMap Map.empty
+ where
```
[the remark this answers is not in the document]
Comment (Unisay) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Cse.hs]: Ok

**Pull request merged**